contracts/citycoin-core.clar

```diff
@@ -1,201 +1,539 @@
+;; or, an empty structure
+(define-read-only (get-stacker-info-at-cycle (rewardCycle uint) (userId uint))
+  (default-to { amountStacked: u0, toReturn: u0 }
+    (map-get? StackersAtCycle { rewardCycle: rewardCycle, userId: userId }))
+)
+
+;; get the reward cycle for a given Stacks block height
+(define-read-only (get-reward-cycle (stacksHeight uint))
   (let
     (
-      (contract (unwrap! (get-contract address) (err ERR_CONTRACT_DOES_NOT_EXIST)))
+      (firstStackingBlock (var-get activationBlock))
+      (rcLen (var-get rewardCycleLength))
     )
-    (map-set ActiveContracts
-      (get name contract)
-      address
-    )   
-    ;; call somehow startup
-    (ok true)
+    (if (>= stacksHeight firstStackingBlock)
+      (some (/ (- stacksHeight firstStackingBlock) rcLen))
+      none)
   )
 )
 
+;; determine if stacking is active in a given cycle
+(define-read-only (stacking-active-at-cycle (rewardCycle uint))
+  (is-some
+    (get amountToken (map-get? StackingStatsAtCycle rewardCycle))
+  )
+)
 
-;; --------------------
-(define-private (is-authorized)
-  (is-eq contract-caller (var-get cityWallet))
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; TOKEN
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; how many blocks until the next halving occurs
+(define-constant TOKEN_HALVING_BLOCKS u210000)
+
+;; store block height at each halving, set by register-user    
+(define-data-var coinbaseThreshold1 uint u0)
+(define-data-var coinbaseThreshold2 uint u0)
+(define-data-var coinbaseThreshold3 uint u0)
+(define-data-var coinbaseThreshold4 uint u0)
+(define-data-var coinbaseThreshold5 uint u0)
+
+;; return coinbase thresholds if contract activated
+(define-read-only (get-coinbase-thresholds)
+  (let
+    (
+      (activated (var-get activationReached))
+    )
+    (asserts! activated (err ERR_CONTRACT_NOT_ACTIVATED))
+    (ok {
+      coinbaseThreshold1: (var-get coinbaseThreshold1),
+      coinbaseThreshold2: (var-get coinbaseThreshold2),
+      coinbaseThreshold3: (var-get coinbaseThreshold3),
+      coinbaseThreshold4: (var-get coinbaseThreshold4),
+      coinbaseThreshold5: (var-get coinbaseThreshold5)
+    })
+  )
 )
 
-;; helper function
-(define-read-only (is-between (value uint) (low uint) (high uint))
-  (and 
-    (>= value low)
-    (<= value high)
+;; function for deciding how many tokens to mint, depending on when they were mined
+(define-read-only (get-coinbase-amount (minerBlockHeight uint))
+  (begin
+    ;; if contract is not active, return 0
+    (asserts! (>= minerBlockHeight (var-get activationBlock)) u0)
+    ;; if contract is active, return based on issuance schedule
+    ;; halvings occur every 210,000 blocks for 1,050,000 Stacks blocks
+    ;; then mining continues indefinitely with 3,125 CityCoins as the reward
+    (asserts! (> minerBlockHeight (var-get coinbaseThreshold1))
+      (if (<= (- minerBlockHeight (var-get activationBlock)) u10000)
```

Comment:
In theory this subtraction can underflow, but it will be captured immediately by Clarity, as we use only unsigned integers.

When it comes to logic of these two lines - they are responsible for handling bonus reward period (first 10000 blocks) for early miners. 
You can find detailed description of issuance schedule here: https://github.com/citycoins/citycoin/blob/main/citycoin-prd.md#issuance-schedule

Reply:
We're definitely using `asserts!` in an interesting pattern here, but it helped alleviate a ton of nested if statements.

Since the `activationBlock` variable is set as the [block when mining can officially begin](https://github.com/citycoins/citycoin/blob/feat/simplify-logic/contracts/citycoin-core.clar#L221), and `coinbaseThreshold1` is set to `activationBlock + TOKEN_HALVING_BLOCKS (u210,000)`, there should not be a way you can reach this code path with an underflow.

We had some pretty extensive testing on the [original version here](https://github.com/citycoins/citycoin/blob/main/tests/citycoin-logic-v1.test.ts#L136), which will be implemented again to make sure all expected values are returned.